Qml Http Server-based storybook

micieslak opened the pull request:
# What does this PR do

This PR introduces possibility to load files via http server instead of directly from the file system. It gives ability to run it on mobile devices with hot-reloading, where server is hosted on a dev machine.

micieslak committed:
cmake config for tests fixed

micieslak committed:
Page status enum decoupled from PagesModel type

micieslak committed:
StorybookData singleton introduced

micieslak committed:
pagesFolder ctx prop moved to StorybookData

micieslak committed:
workaround for QTBUG-111789

micieslak committed:
QmlFilesServer introduced

micieslak committed:
AbstractPagesSource introduced

micieslak committed:
move paths-related logic from PagesModel to LocalPagesSource

micieslak committed:
RemotePagesSource added for reading pages list/metadata via rest api

micieslak committed:
QmlFilesServer files version tracking added

micieslak committed:
CacheCleaner replaced with QmlEngineUtils

micieslak committed:
Changes notifier interface introduced

micieslak committed:
RemoteChangesNotifier added

micieslak committed:
Mode enum exposed to qml

micieslak committed:
Add top-level config for remote setup

micieslak committed:
Page loading adjusted to both local and remote mode

micieslak committed:
QmlFilesServer: added ability to amend qmldirs on the fly

When accessing source files via http, only files listed in 'qmldir's can
be accessed. When loading from filesystem or qrc, there is no such
limitation, therefore some files are not listed in their corresponding
qmldir file. In order to seamlessly load sources via http even if some
files are not listed in qmldir, this commit adds ability of amending
qmldir files on the fly, adding missing qml files.

caybro commented on src/qmlfilesserver.cpp:
```diff
@@ -0,0 +1,229 @@
+    QString singletonKeyword;
+    QSet<QString> listedFiles;
+
+    for (const QString& line : qmldirLines) {
```
```suggestion
    for (const QString& line : std::as_conts(qmldirLines)) {
```

caybro commented on src/qmlfilesserver.cpp:
```diff
@@ -0,0 +1,229 @@
+
+        QDir dir(m_pagesPath);
+        dir.setFilter(QDir::Files);
+        dir.setNameFilters({"*Page.qml"});
```
```suggestion
        dir.setNameFilters({QStringLiteral("*Page.qml")});
```

caybro commented on src/qmlfilesserver.cpp:
```diff
@@ -0,0 +1,128 @@
+        const QFileInfoList filesInfo = dir.entryInfoList();
+
+        QJsonArray files;
+        std::transform(filesInfo.begin(), filesInfo.end(),
+                       std::back_inserter(files),
+                       [] (auto &info) {
+
+            return QJsonObject {
+                {
+                    { u"page"_s, info.fileName() },
+                    { u"timestamp"_s, info.lastModified().toMSecsSinceEpoch() }
+                }
+            };
+        });
+
```
Actually a nice suggestion :)

micieslak replied:
This suggestion makes no sense, there is `push_back` ofc, otherwise it wouldn't compile...

caybro replied:
Yeah true, I was just saying it's nicer code :)

caybro commented on src/qmlfilesserver.cpp:
```diff
@@ -0,0 +1,229 @@
+
+        QFileInfo info(relativePath);
+
+        if (info.fileName() == "qmldir") {
```
```suggestion
        if (info.fileName() == QStringLiteral("qmldir")) {
```

caybro commented on src/qmlfilesserver.cpp:
```diff
@@ -0,0 +1,229 @@
+
+QString QmlFilesServer::findFirstExistingSourceDir(const QString &relativePath) const
+{
+    for (const QString& base : m_basePaths)
```
```suggestion
    for (const QString& base : std::as_const(m_basePaths))
```

micieslak replied:
hmm, but we are in `const` method, `std::as_const` will do nothing there as all members are const already...

caybro commented on src/qmlfilesserver.cpp:
```diff
@@ -0,0 +1,229 @@
+
+QString QmlFilesServer::findFirstExistingFile(const QString &relativePath) const
+{
+    for (const QString& base : m_basePaths)
```
```suggestion
    for (const QString& base : std::as_const(m_basePaths))
```

caybro commented on src/remotepagessource.cpp:
```diff
@@ -0,0 +1,179 @@
+QStringList RemotePagesSource::pages() const
+{
+    QStringList pages;
+    for (auto& [page, _] : m_pages)
```
```suggestion
    for (auto& [page, _] : std::as_const(m_pages))
```

caybro commented on src/remotepagessource.cpp:
```diff
@@ -0,0 +1,179 @@
+    auto it = m_pagesContent.find(key);
+
+    if (it == m_pagesContent.cend())
+        return "";
```
```suggestion
        return {};
```

caybro commented on src/remotepagessource.cpp:
```diff
@@ -0,0 +1,179 @@
+                    obj.value("timestamp").toInteger());
+        }
+
+        for (const auto& [page, timestamp] : m_pagesNew) {
```
```suggestion
        for (const auto& [page, timestamp] : std::as_const(m_pagesNew)) {
```

caybro commented on src/remotepagessource.cpp:
```diff
@@ -0,0 +1,179 @@
+                m_pagesContentNew[page] = m_pagesContent[page];
+        }
+
+        for (const auto& [page, _] : m_pages)
```
```suggestion
        for (const auto& [page, _] : std::as_const(m_pages))
```

caybro reviewed (commented):
Nice work, just some minor remarks

micieslak committed:
Hot reloading: remote loading no longer relies on clearComponentCache

clearComponentCache was causing crashes in some cases. Now the reload is
enforced by changing import url whenever source code change is detected.

micieslak committed:
Unused components related to hot-reloading removed

micieslak committed:
Irrelevant controls hidden when run on mobile

micieslak committed:
SwipeView-based layout added for mobile

caybro commented on qml/Storybook/Storybook.qml:
```diff
@@ -11,6 +11,9 @@ Control {
     readonly property alias currentPage: d.currentPage
     readonly property alias darkMode: darkModeCheckBox.checked
 
+    readonly property bool mobile: true || Qt.platform.os === "ios" ||
```
Uhm :D

alexjba reviewed (approved):
Awesome work! 🚀

caybro reviewed (approved):
Noice!
